ek9000App/src/scripts/generate.py

```diff
@@ -183,35 +196,110 @@ def write(self, fp):
                 fp.write(f'\tfield({i[0]}, "{i[1]}")\n')
             fp.write('}\n\n')
 
+
+    @staticmethod
+    def _get_raw_range(spec: SpecType) -> Tuple[int,int]:
+        """
+        Computes the raw range for the terminal
+        
+        Returns
+        -------
+        Tuple[int,int]:
+            Range interval (min,max)
+        """
+        isSigned = False
+        match spec['representation']:
+            case 'int16':
+                isSigned = True
+            case 'uint16':
+                isSigned = False
+            case other:
+                assert False
+        max = 32767
+        min = -max if spec['min'] < 0 and isSigned else 0 # Unsigned terminals still have range 0-32767
+        return (min, max)
+
+
+    @staticmethod
+    def _compute_deadbands(spec: SpecType) -> float:
+        """
+        Compute deadband parameters ADEL and MDEL based on max EGU and error specified in datasheet
+        Most terminals are rated for measurement error of <=0.3%
+        
+        Returns
+        -------
+        float:
+            The value ADEL and MDEL should be set to
+        """
+        error = spec['error'] / 100.0 # error is in %, convert that to float
+        return spec['max'] * error # Docs specify that error is relative to the full scale value
+
+
+    def _compute_analog_defaults(self) -> dict:
+        """
+        Compute ESLO, EOFF, EGUF and EGUL for the analog I/O terminal
+        
+        Returns
+        -------
+        dict:
+            Mapping of PV -> values
+        """
+
+        if self.spec is None:
+            return {}
+
+        rawMin, rawMax = self._get_raw_range(self.spec)
+        range = self.spec['max'] - self.spec['min']
+        deadband = self._compute_deadbands(self.spec)
+        return {
+            'ESLO': (range / (rawMax-rawMin)),
+            'EOFF': (rawMax * self.spec['min'] - rawMin * self.spec['max']) / (rawMax-rawMin),
```

Comment:
I'm trying to get to the bottom of this one - correct me if I'm wrong on any of the stuff below. That said with the comment below regarding `LINR`, it might not be necessary to calculate this at all?

Running with a specific example, the EL3051 which has a signed 16-bit representation (12 bit resolution), the spec file shows 4-20mA with 0.3% error. The docs indicate:
<img width="872" alt="image" src="https://github.com/slaclab/epics-ek9000/assets/5139267/48d98ef3-8177-4939-9a1d-3ca84f241022">

In the code, raw min with signed int16 storage is 0, raw max is 0x7FFF - which matches up with 4mA at 0 and 20mA at 0x7FFF. Then EOFF gets set to `(0x7fff * 4 - 0 * 20) / 0x7fff` or `4.0` (which makes sense because we offset a value of 0 up to 4mA in this scenario). That seems like a roundabout way to get to "the minimum value of 4.0mA" - can you describe how you derived the above?

It also looks like you can reconfigure it to also use signed/unsigned/MSB absolute:
<img width="894" alt="image" src="https://github.com/slaclab/epics-ek9000/assets/5139267/5bec418f-c190-4946-b8bc-bcf79d66623a">

If we generate the records in advance, we can't account for that unfortunately. For unipolar (positive values only), the representation won't change. I think it's fine to not support bipolar-capable terminals that are (erroneously) configured with a different presentation. It might be worth a note somewhere (use standard PDO mapping, default presentation).

Reply:
Oh, maybe this came directly from the EPICS docs?

```
    EOFF = (RVAL_max * EGUL − RVAL_min * EGUF) / (RVAL_max − RVAL_min)

    ESLO = (EGUF − EGUL) / (RVAL_max − RVAL_min)
```

Alright, this is probably good but if you could walk me through the details I'd feel a bit better.

Reply:
Yes, the `EOFF` formula came from the EPICS docs, I should've probably left a comment indicating that. 

In my testing it yields the correct `EOFF` for both unipolar and bipolar terminals using the values returned from `_get_raw_range`. However I made a mistake in `_get_raw_range`: min != -max, it should be `-max-1` (-32768). So that formula isn't exactly correct for bipolar terminals. With that formula `EOFF` is computed to be ~0.00015 with min=-32768 and max=32767, which makes 0 ADC counts be just above 0 EGU. So, I'm just going to simplify this to be `0` for bipolar terminals, and `spec['min']` for unipolar.

You're right about the `LINR` setting, it should be `SLOPE` rather than `LINEAR`.

Reply:
> If we generate the records in advance, we can't account for that unfortunately. For unipolar (positive values only), the representation won't change. I think it's fine to not support bipolar-capable terminals that are (erroneously) configured with a different presentation. It might be worth a note somewhere (use standard PDO mapping, default presentation).

I'll do a documentation pass soon and get ReadTheDocs setup. There'll be a section on commissioning and some other stuff described in #42